File: proyecto1/torreControl.cpp
#include "torreControl.h"

TorreControl::TorreControl()
{
    PuertaEmbarque puerta1(1);
    PuertaEmbarque puerta2(2);
    PuertaEmbarque puerta3(3);
    PuertaEmbarque puerta4(4);
    PuertaEmbarque puerta5(5);
    PuertaEmbarque puerta6(6);

    puertas.push_back(puerta1);
    puertas.push_back(puerta2);
    puertas.push_back(puerta3);
    puertas.push_back(puerta4);
    puertas.push_back(puerta5);
    puertas.push_back(puerta6);
}

void TorreControl::registrarAeronave(Aeronave *aeronave)
{
    aeronaves.push_back(aeronave);
}
void TorreControl::registrarAvion(Aeronave *aeronave)
{
    aviones.push_back(aeronave);
}
/**
 * The function "enviarMensaje" in the "TorreControl" class sends a message to all aircrafts except the
 * one that is the sender.
 *
 * The mensaje parameter is a constant reference to a string, which represents the
 * message that will be sent to the aeronaves.
 * The parameter "emisor" is a pointer to an object of type "Aeronave". It represents the
 * aircraft that is sending the message.
 */
void TorreControl::enviarMensaje(const string &mensaje, Aeronave *emisor)
{
    for (auto &aeronave : aeronaves)
    {
        if (aeronave != emisor)
        {
            aeronave->recibirMensaje(mensaje);
        }
    }
}

void TorreControl::asignarPuertaDeEmbarque(Aeronave *aeronave, int puerta)
{
    aeronave->asignarPuertaDeEmbarque(puerta);
    puertas[puerta - 1].disponibilidad = false;
}

bool TorreControl::disponibilidadNaves()
{
    if (aeronaves.size() == 0)
        return false;
    else
        return true;
}

void TorreControl::mostrarAviones()
{

    for (int i = 0; i < aeronaves.size(); i++)
    {
        printf("%d.\n", i + 1);
        aeronaves[i]->printInfo();
    }
}

void TorreControl::seleccionarAeronave(Vuelos *v)
{
    for (auto &aeronave : aeronaves)
    {
        if (aeronave->estado)
        {
            aeronave->agregarVuelo(v);
            bool flag = true;
            for (int i = 0; i < puertas.size() && flag; i++)
            {
                if (puertas[i].disponibilidad)
                {
                    this->asignarPuertaDeEmbarque(aeronave, puertas[i].identificacion);
                    flag = false;
                }
            }
            break;
        }
    }
}

int generarNumeroAleatorio()
{
    // Inicializa la semilla del generador de números aleatorios
    std::srand(static_cast<unsigned int>(std::time(nullptr)));

    int num1 = rand() % 90001 + 10000;
    int num2 = rand() % 90001 + 10000;
    // Genera un número aleatorio entre 10,000 y 100,000
    int numeroAleatorio = std::rand() % num1 + num2; // [0, 90000] + 10000 = [10000, 100000]

    return numeroAleatorio;
}

void TorreControl::simulacion()
{
    int pos1, pos2;

    for (auto &aeronave : aeronaves)
    {
        if (aeronave->tieneVuelos())
        {
            aeronave->despegar();
            pos1 = generarNumeroAleatorio();
            pos2 = generarNumeroAleatorio();
            string n = "Lat: ";
            string m = " Lon: ";
            string tmp = to_string(pos1);
            n += tmp;
            n += m;
            tmp = to_string(pos2);
            n += tmp;
            aeronave->actualizarPosicion(n);
        }
    }
    for (auto &aeronave : aeronaves)
    {
        if (aeronave->tieneVuelos())
        {
            aeronave->aterrizar();
            aeronave->eliminarVuelo();
        }
    }
    for (int i = 0; i < puertas.size(); i++)
    {
        puertas[i].disponibilidad = true;
    }
}

File: proyecto2/torreControl.py
from mediadorTrafico import MediadorTrafico
from puertaEmbarque import PuertaEmbarque
from aeronave import Aeronave
from vuelos import Vuelos
import streamlit as st


class TorreControl(MediadorTrafico):
    def __init__(self):
        self.aeronaves = []
        self.aviones = []
        self.puertas = [PuertaEmbarque(i) for i in range(1, 7)]
        self.cont = 0

    def registrarAeronave(self, aeronave: Aeronave):
        self.aeronaves.append(aeronave)

    def registrarAvion(self, aeronave: Aeronave):
        self.aviones.append(aeronave)

    def enviarMensaje(self, mensaje, emisor: Aeronave):
        s = ""
        for aeronave in self.aeronaves:
            if aeronave != emisor:
                s += aeronave.recibirMensaje(mensaje) + ","

        print(s)
        return s

    def asignarPuertaDeEmbarque(self, aeronave: Aeronave, puerta, cod, hora):
        aeronave.asignarPuertaDeEmbarque(puerta, cod, hora)
        self.puertas[puerta - 1].disponibilidad = False

    def disponibilidadNaves(self):
        return bool(self.aeronaves)

    def mostrarAeronaves(self):
        x=[]
        for i in range(len(self.aeronaves)):
            x.append(self.aeronaves[i].printInfo())
        return x

    def seleccionarAeronave(self, vuelo : Vuelos):
        if self.cont == len(self.aeronaves):
            self.cont = 0
        for j in range(self.cont, len(self.aeronaves)):
            if self.aeronaves[j].estado:
                self.aeronaves[j].agregarVuelo(vuelo)
                flag = True
                for i in range(len(self.puertas)):
                    if self.puertas[i].disponibilidad:
                        self.asignarPuertaDeEmbarque(self.aeronaves[j], self.puertas[i].ident, vuelo.identificacion, vuelo.hora)
                        flag = False
                        break
                break
        self.cont += 1

    def simulacion(self):
        def generarNumeroAleatorio():
            import random
            return random.randint(10000, 100000)
        lista2=[]

        for aeronave in self.aeronaves:
            
            if aeronave.tieneVuelos():
                for i in range(len(aeronave.vuelos)):
                    x=aeronave.despegar()
                    pos1 = generarNumeroAleatorio()
                    pos2 = generarNumeroAleatorio()
                    n = "Lat: " + str(pos1) + " Lon: " + str(pos2)
                    y=aeronave.actualizarPosicion(n)
                    z=aeronave.aterrizar()
                    lista=[x,y,z]
                    lista2.append(lista)

                for i in range(len(aeronave.vuelos)):
                    aeronave.eliminarVuelo()
        

        for puerta in self.puertas:
            puerta.disponibilidad = True
        return lista2

    def mostrarPuertas(self):
        l=[]
        for puerta in self.puertas:
            estado = "disponible" if puerta.disponibilidad else "no disponible"
            x=(f"Puerta #{puerta.identificacion} {estado}")
            l.append(x)
        return l
    
    def listPuertas(self):
        l = []
        for i in range(len(self.puertas)):
            l.append(self.puertas[i].to_dictPE())
        return l
